### PR TITLE
esp8266/modpybuart: allow setting UART baudrate and other params

### PR DESCRIPTION
**Pull request by deshipu**

### CODE REVIEW

**Comment by deshipu** [comment]
Parity should probably be set using constants, but I would need some help with that.

File: esp8266/modpybuart.c [hunk 1]
```diff
@@ -70,16 +75,114 @@ STATIC void pyb_uart_init_helper(pyb_uart_obj_t *self, size_t n_args, const mp_o
...
+        case 74880:
+        case 115200:
+        case 230400:
+        case 256000:
+        case 460800:
+        case 921600:
+        case 1843200:
+        case 3686400:
+            self->baudrate = args[ARG_baudrate].u_int;
+            UartDev.baut_rate = self->baudrate; // Sic!
```
Comment (dpgeorge) [inline esp8266/modpybuart.c]: Is it really restricted to these baudrates?  Why not just let anything through?  Or at least a range of values that yield valid baudrates.

**Comment by dpgeorge** [comment]
> Parity should probably be set using constants

On pyboard it's an odd number for odd parity, an even number for even parity, and `None` for no parity.

File: esp8266/modpybuart.c (hunk 1 above)
Comment (deshipu) [inline esp8266/modpybuart.c]: I don't know. I mostly based this on how NodeMCU does it, and it restricts it to those baud rates. I will experiment a little and see if I can get some weird baud rates out of this.

Comment (dhylands) [inline esp8266/modpybuart.c]: Some odd baud rates I've used/seen: 31250 for MIDI, 250,000, 500,000, and 1,000,000 for talking to bioloid servos.

**Comment by deshipu** [comment]
It seems that the esp supports arbitrary baud rates, although not all of them are supported by the usb2ttl on the HUZZAH board. I also switched to 115200 baud by default (since that's what the uart starts with), and use the uart_id in the setup call (even though it's always 0 anyways).

How would I go about recognizing None as the parity value?

**Comment by dpgeorge** [comment]
> How would I go about recognizing None as the parity value?

Look at how stmhal/uart.c does it.

**Comment by deshipu** [comment]
Hmm, OK, but that sets the parity to the default None if it's not specified. I'm trying to make it only change the parameters that are specified. What would be a good choice for the default value that is not None here?

**Comment by deshipu** [comment]
So I have parity covered (used False as the default value, not pretty, but works). Is there anything else that needs attention here, or should I squash and rebase?

File: esp8266/modpybuart.c [hunk 2]
```diff
@@ -36,32 +36,40 @@
...
+// UartDev is defined and initialized in rom code.
+extern UartDevice UartDev;
 
 typedef struct _pyb_uart_obj_t {
     mp_obj_base_t base;
     uint8_t uart_id;
+    uint32_t baudrate;
+    uint8_t bits;
+    uint8_t parity;
+    uint8_t stop;
```
Comment (dpgeorge) [inline esp8266/modpybuart.c]: Please reorder so the 4 uint8_t's are together to improved the packing of the struct.

File: esp8266/modpybuart.c [hunk 3]
```diff
@@ -70,16 +78,87 @@ STATIC void pyb_uart_init_helper(pyb_uart_obj_t *self, size_t n_args, const mp_o
...
+            UartDev.data_bits = UART_SEVEN_BITS;
+            self->bits = 7;
+            break;
+        case 8:
+            UartDev.data_bits = UART_EIGHT_BITS;
+            self->bits = 8;
+            break;
+        default:
+            nlr_raise(mp_obj_new_exception_msg_varg(
+                &mp_type_ValueError, "Invalid data bits"));
```
Comment (dpgeorge) [inline esp8266/modpybuart.c]: Error messages should start with a lower case letter.

File: esp8266/modpybuart.c [hunk 4]
```diff
@@ -70,16 +78,87 @@ STATIC void pyb_uart_init_helper(pyb_uart_obj_t *self, size_t n_args, const mp_o
...
+            UartDev.stop_bits = UART_ONE_STOP_BIT;
+            self->stop = 1;
+            break;
+        case 2:
+            UartDev.stop_bits = UART_TWO_STOP_BIT;
+            self->stop = 2;
+            break;
+        default:
+            nlr_raise(mp_obj_new_exception_msg_varg(
+                &mp_type_ValueError, "Invalid stop bits"));
```
Comment (dpgeorge) [inline esp8266/modpybuart.c]: Lower case.

File: esp8266/modpybuart.c (hunk 3 above)
Comment (dpgeorge) [inline esp8266/modpybuart.c]: Also, put it on one line if its within 90-ish characters.

File: esp8266/uart.h [hunk 5]
```diff
@@ -96,5 +96,7 @@ int uart_rx_char(void);
 void uart_tx_one_char(uint8 uart, uint8 TxChar);
 void uart_flush(uint8 uart);
 void uart_os_config(int uart);
+void ICACHE_FLASH_ATTR uart_setup(uint8 uart_no);
```
Comment (dpgeorge) [inline esp8266/uart.h]: You don't need the ICACHE_FLASH_ATTR here.

File: esp8266/uart.h [hunk 6]
```diff
@@ -96,5 +96,7 @@ int uart_rx_char(void);
 void uart_tx_one_char(uint8 uart, uint8 TxChar);
 void uart_flush(uint8 uart);
 void uart_os_config(int uart);
+void uart_setup(uint8 uart_no);
+
```
Comment (dpgeorge) [inline esp8266/uart.h]: Unnecessary blank.

File: esp8266/modpybuart.c [hunk 7]
```diff
@@ -70,16 +78,85 @@ STATIC void pyb_uart_init_helper(pyb_uart_obj_t *self, size_t n_args, const mp_o
...
+
+    // set data bits
+    switch (args[ARG_bits].u_int) {
+        case 0:
+            break;
+        case 5:
+            UartDev.data_bits = UART_FIVE_BITS;
+            self->bits = 5;
+            break;
+
```
Comment (dpgeorge) [inline esp8266/modpybuart.c]: Unnecessary blank line.

File: esp8266/modpybuart.c [hunk 8]
```diff
@@ -36,32 +36,40 @@
...
+    uint8_t parity;
+    uint8_t stop;
+    uint32_t baudrate;
     uint16_t timeout;       // timeout waiting for first char (in ms)
     uint16_t timeout_char;  // timeout waiting between chars (in ms)
 } pyb_uart_obj_t;
 
+STATIC const char *_parity_name[] = {"None", "1", "0"};
+
+
```
Comment (dpgeorge) [inline esp8266/modpybuart.c]: Unnecessary blank line.

**Comment by dpgeorge** [comment]
Please squash after fixing blank lines.

**Comment by deshipu** [comment]
Corrected and squashed, thanks!

**Comment by pfalcon** [comment]
> used False as the default value, not pretty, but works

The obvious choice for "unset" value is MP_OBJ_NULL, if you grep for ".u_obj =", you'll see that it's used in a lot of places.

File: esp8266/uart.c [hunk 9]
```diff
@@ -237,6 +237,13 @@ void ICACHE_FLASH_ATTR uart_reattach() {
     uart_init(UART_BIT_RATE_74880, UART_BIT_RATE_74880);
 }
 
+void ICACHE_FLASH_ATTR uart_setup(uint8 uart_no) {
+    ETS_UART_INTR_DISABLE();
+    uart_config(uart_no);
+    ETS_UART_INTR_ENABLE();
+}
+
+
```
Comment (pfalcon) [inline esp8266/uart.c]: There's still extra blank line.

File: esp8266/uart.h [hunk 10]
```diff
@@ -96,5 +96,6 @@ int uart_rx_char(void);
 void uart_tx_one_char(uint8 uart, uint8 TxChar);
 void uart_flush(uint8 uart);
 void uart_os_config(int uart);
+void uart_setup(uint8 uart_no);
```
Comment (pfalcon) [inline esp8266/uart.h]: Minor, but as you can see above, this argument name is called "uart" for the other functions, and consistency should rather be preserved.

**Comment by deshipu** [comment]
Removed one more empty, line, used ``MP_OBJ_NULL`` and renamed ``uart_id`` to ``uart``.

**Comment by pfalcon** [comment]
Perfect, thanks! Merged.

**Comment by Himura2la** [comment]
Hi guys, can you please give me a clue on how to change the MicroPython REPL's baud rate on ESP826? I found no docs for it...

**Comment by Himura2la** [comment]
For those who came here from Google in search of a way to change the REPL's baud rate, here's the link 
http://docs.micropython.org/en/latest/esp8266/library/machine.UART.html

**Pull request merged**